src/main/java/sg/ncl/MainController.java

```diff
@@ -2105,113 +2101,118 @@ public String data(Model model, HttpSession session) throws Exception {
-//    public String editDatasetInfo(@PathVariable Integer datasetId, @ModelAttribute("editDataset") Dataset dataset, final RedirectAttributes redirectAttributes) {
-//    	Dataset origDataset = datasetManager.getDataset(datasetId);
-//
-//    	String editedDatasetName = dataset.getDatasetName();
-//    	String editedDatasetDesc = dataset.getDatasetDescription();
-//    	String editedDatasetLicense = dataset.getLicense();
-//    	String editedDatasetPublic = dataset.getIsPublic();
-//    	boolean editedDatasetIsRequiredAuthorization = dataset.getRequireAuthorization();
-//
-//    	System.out.println(origDataset.getDatasetId());
-//    	System.out.println(dataset.getDatasetId());
-//
-//    	if (origDataset.updateName(editedDatasetName) == true) {
-//    		redirectAttributes.addFlashAttribute("editName", "success");
-//    	}
-//
-//    	if (origDataset.updateDescription(editedDatasetDesc) == true) {
-//    		redirectAttributes.addFlashAttribute("editDesc", "success");
-//    	}
-//
-//    	if (origDataset.updateLicense(editedDatasetLicense) == true) {
-//    		redirectAttributes.addFlashAttribute("editLicense", "success");
-//    	}
-//
-//    	if (origDataset.updatePublic(editedDatasetPublic) == true) {
-//    		redirectAttributes.addFlashAttribute("editPublic", "success");
-//    	}
-//
-//    	if (origDataset.updateAuthorization(editedDatasetIsRequiredAuthorization) == true) {
-//    		redirectAttributes.addFlashAttribute("editIsRequiredAuthorization", "success");
-//    	}
-//
-//    	datasetManager.updateDatasetDetails(origDataset);
-//
-//    	return "redirect:/data/edit/{datasetId}";
-//    }
+        try {
+            if (RestUtil.isError(response.getStatusCode())) {
+                MyErrorResource error = objectMapper.readValue(dataResponseBody, MyErrorResource.class);
+                ExceptionState exceptionState = ExceptionState.parseExceptionState(error.getError());
 
-//    @RequestMapping("/data/remove_dataset/{datasetId}")
-//    public String removeDataset(@PathVariable Integer datasetId) {
-//    	datasetManager.removeDataset(datasetId);
-//    	return "redirect:/data";
-//    }
+                switch (exceptionState) {
+                    case DATASET_NAME_IN_USE_EXCEPTION:
+                        model.addAttribute("message", "Error(s):<ul><li>dataset name already exists</li></ul>");
+                        break;
+                    case FORBIDDEN_EXCEPTION:
+                        model.addAttribute("message", "Error(s):<ul><li>saving dataset forbidden</li></ul>");
+                        break;
+                }
+
+                return "data_contribute";
+            }
+        } catch (IOException e) {
+            throw new WebServiceRuntimeException(e.getMessage());
+        }
+
+        return "redirect:/data";
+    }
+
+    @RequestMapping("/data/remove/{id}")
+    public String removeDataset(@PathVariable String id, RedirectAttributes redirectAttributes) throws Exception {
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Define and throw a dedicated exception instead of using a generic one. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://sonarqube.com/coding_rules#rule_key=squid%3AS00112)